[RyzenAI] Add YOLO and TIMM pipelines for Ryzen

mht-sharma opened the pull request:
Support pipeline for the Ryzen models

* [x] Supports transformers for image classification
* [x] Support Timm
* [ ] Object Detection
  * [x] YoloX
  * [x] Yolov5
  * [x] Yolov3
  * [x] Yolov8
* [ ] Code cleaning
* [x] Tests
* [x] Docs

Usage:
```diff
-from transformers import pipeline
+from optimum.amd.ryzenai import pipeline
```

Examples
* Usage `Timm`  models:

```python
import requests
from PIL import Image

from optimum.amd.ryzenai import pipeline


url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

model_id = "mohitsha/timm-resnet18-onnx-quantized-ryzen"

pipe = pipeline("image-classification", model=model_id, vaip_config="vaip_config.json")
print(pipe(image))
```
or

```python
import requests
from PIL import Image

from optimum.amd.ryzenai import RyzenAIModelForImageClassification
from optimum.amd.ryzenai import pipeline


url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

model_id = "mohitsha/timm-resnet18-onnx-quantized-ryzen"
model = RyzenAIModelForImageClassification.from_pretrained(model_id, vaip_config="vaip_config.json")

pipe = pipeline("image-classification", model=model)
print(pipe(image))
```
Usage `Yolo`  models:

```python
import requests
from PIL import Image

from optimum.amd.ryzenai import pipeline

img = ".\\image.jpg"
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
img = ".\\image2.jpg"

image = Image.open(img)

# model_id = "amd/yolox-s"
# model_id = "amd/yolov3"
# model_id = "amd/yolov5s"
model_id = "amd/yolov8m"

pipe = pipeline("object-detection", model=model_id, vaip_config="vaip_config.json", model_type="yolov8")
pipe(image)
```

fxmarty commented on .github/workflows/build_pr_documentation.yml:
```diff
@@ -53,7 +53,7 @@ jobs:
           pip install .
 
-          pip install onnxruntime==1.14.0
+          pip install onnxruntime==1.14.0 torchvision
```
maybe edit the setup.py as well if this is required

mht-sharma replied:
done

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
```diff
@@ -0,0 +1,123 @@
+
+Currently the supported tasks are:
+
+* `image-classification`
+* `object-detection` (Supported model types - yolox, yolov3, yolov5, yolov8)
```
You could link to the relevant Hub models (with filters)

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
```diff
@@ -0,0 +1,123 @@
+For more information refer to [runtime-configuration-file](https://ryzenai.docs.amd.com/en/latest/runtime_setup.html#runtime-configuration-file)
+</Tip>
+
+## The pipeline abstraction
+The pipeline abstraction is a wrapper around all the available pipelines for specific tasks. The pipeline() function automatically loads a default model and tokenizer/feature-extractor capable of performing inference for your task.
```
You could link to Transformers documentation about pipelines, or mention that this is derived from it

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
````diff
@@ -0,0 +1,123 @@
+```
+
+<Tip warning={true}>
+The detault model is only supported for `object-detection task. For other tasks, please provide the `model_id / RyzenAIModel` for inference.`
````
```suggestion
The default model is only supported for the `object-detection` task. For other tasks, please provide a model name (local path or model hosted on the Hub) or a `RyzenAIModel` for inference.`
```
tbh could be removed as well

mht-sharma replied:
removed

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
```diff
@@ -0,0 +1,123 @@
+
+## Using Ryzen models
+
+The [`~ryzenai.pipelines.pipeline`] function accepts Ryzen supported models from the [Hugging Face Hub](https://huggingface.co/models?other=RyzenAI)
```
```suggestion
The [`~ryzenai.pipelines.pipeline`] function can load Ryzen AI supported models from the [Hugging Face Hub](https://huggingface.co/models?other=RyzenAI).
```

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
```diff
@@ -0,0 +1,123 @@
+
+The [`~ryzenai.pipelines.pipeline`] function accepts Ryzen supported models from the [Hugging Face Hub](https://huggingface.co/models?other=RyzenAI)
+
+Once you have picked an appropriate model, you can create the [`~ryzenai.pipelines.pipeline`] by specifying the model repo:
```
```suggestion
Once you have picked an appropriate model, you can create the [`~ryzenai.pipelines.pipeline`] by specifying the model repository:
```

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
````diff
@@ -0,0 +1,123 @@
+print(pipe(image))
+```
+
+It is also possible to load it with the `from_pretrained(model_name_or_path, export=True)`
````
How does this work? Where is quantization done?

mht-sharma replied:
The model  in the `model_id` is already quantized

fxmarty commented on docs/source/ryzenai/usage_guides/pipelines.mdx:
```diff
@@ -0,0 +1,123 @@
+
+Note for a few models, `model_type` amd `image_preprocessor` has to be provided in addition to the model_id for inference.
+
+For example, here is how you can run inference using `yolox`
```
```suggestion
For example, here is how you can run inference using `yolox`:
```

fxmarty commented on optimum/amd/ryzenai/models/detection_utils.py:
```diff
@@ -0,0 +1,116 @@
+    class_conf_start_index=5,
+    agnostic=False,
+    max_detections=300,
+):
+    output = [torch.zeros((0, 6), device=predictions.device)] * predictions.shape[0]
+
+    for image_idx, _ in enumerate(predictions):
+        prediction = predictions[image_idx]
+
+        prediction = prediction[has_confidence[image_idx]]
+
+        if not prediction.shape[0]:
+            continue
+
+        boxes = center_to_corners_format(prediction[:, :4])
+
+        if class_conf_start_index == 5:
+            prediction[:, 5:] *= prediction[:, 4:5]
+
+        confidence_values, class_indices = prediction[:, class_conf_start_index:].max(1, keepdim=True)
+        prediction = torch.cat((boxes, confidence_values, class_indices.float()), 1)[
+            confidence_values.view(-1) > confidence_threshold
+        ]
+        prediction = prediction[prediction[:, 4].argsort(descending=True)]
+
+        boxes = prediction[:, :4]
+        scores = prediction[:, 4]
+        class_indices = prediction[:, 5]
+
+        if agnostic:
+            nms_indices = torchvision.ops.nms(boxes, scores, iou_threshold)
+        else:
+            nms_indices = torchvision.ops.batched_nms(boxes, scores, class_indices, iou_threshold)
+
+        if nms_indices.shape[0] > max_detections:
+            nms_indices = nms_indices[:max_detections]
+
+        output[image_idx] = prediction[nms_indices]
+
+    return output
```
Could we import them from transformers instead?

mht-sharma replied:
This is not available in transformers

fxmarty replied:
Why can't they not be imported?

mht-sharma replied:
The iou methods are locally implemented in the modeling code for the models. For example: https://github.com/huggingface/transformers/blob/main/src/transformers/models/detr/modeling_detr.py#L2368

To avoid creating unnecessary dependencies on the modeling code, a clean copy of the IoU methods is included here.

fxmarty commented on optimum/amd/ryzenai/models/yolov3/image_processing_yolov3.py:
```diff
@@ -0,0 +1,180 @@
+    return torch.cat(outputs, 1)
+
+
+class YoloV3ImageProcessor(BaseImageProcessor):
```
Why do we need this here? In optimum-amd? Is it not already in transformers?

mht-sharma replied:
Most of the CV models (including yolo and other image2image models) from Ryzen `model_zoo` are not present in transformers. So added these here!

mht-sharma replied:
Unfortunately, none of the HF libraries have these models or similar detection utilities. So, I had to include them here. However, I did receive help from the sample code provided by AMD.

fxmarty commented on optimum/amd/ryzenai/models/yolov5/image_processing_yolov5.py:
```diff
@@ -0,0 +1,183 @@
+    return torch.cat(outputs, 1)
+
+
+class YoloV5ImageProcessor(BaseImageProcessor):
```
same question

fxmarty commented on optimum/amd/ryzenai/models/yolov8/image_processing_yolov8.py:
```diff
@@ -0,0 +1,191 @@
+    return y
+
+
+class YoloV8ImageProcessor(BaseImageProcessor):
```
same

fxmarty commented on optimum/amd/ryzenai/models/yolox/image_processing_yolox.py:
```diff
@@ -0,0 +1,208 @@
+    return outputs
+
+
+class YoloXImageProcessor(BaseImageProcessor):
```
same

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+}
+
+RYZENAI_SUPPORTED_TASKS = {
+    "image-classification": {
+        "impl": ImageClassificationPipeline,
+        "class": (RyzenAIModelForImageClassification,),
+        "default": "amd/resnet50",
+        "type": "image",
+    },
```
pipelines.mdx says there's no default?

mht-sharma replied:
This default model `amd/resnet50` does not work for now without passing the correct `image-processor`. The user needs to pass the `image_processor` with `data_format NHWC` for most of Ryzen pretrained models. 
For object detection, I have created preprocessors so I have taken care of it.

fxmarty replied:
It appears you updated pipelines.mdx accordingly.

mht-sharma replied:
I do throw an error when user tries to create a default model for the task. Possibly will soon update with a model to use for default

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+
+
+def pipeline(
+    task,
+    vaip_config,
```
typing

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+    The pipeline includes components such as a image processor and model.
+
+    Args:
+        task (str):
+            The task defining which pipeline will be returned. Available tasks include:
+            - "image-classification"
+            - "object-detection"
+
```
In the docstring, you need to put types in tics: `str`

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+            Runtime configuration file for inference with Ryzen IPU. A default config file can be found in the Ryzen AI VOE package,
+            extracted during installation under the name `vaip_config.json`.
+
+        model (Optional[Any], *optional*):
```
We decided not to use `*optional*` in optimum libs, let's stick to it https://huggingface.slack.com/archives/C02P0559X9S/p1669628754896019

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+            extracted during installation under the name `vaip_config.json`.
+
+        model (Optional[Any], *optional*):
+            The model that will be used by the pipeline to make predictions. This can be a model identifier or an
+            actual instance of a pretrained model. If not provided, the default model for the specified task will be loaded.
+
+        model_type (Optional[str], *optional*):
+            Model type for the model
+
+        feature_extractor (Union[str, "PreTrainedFeatureExtractor"], *optional*):
+            The feature extractor that will be used by the pipeline to encode data for the model. This can be a model
+            identifier or an actual pretrained feature extractor.
+
+        image_processor (Union[str, BaseImageProcessor], *optional*):
+            The image processor that will be used by the pipeline for image-related tasks.
+
+        use_fast (bool, *optional*, defaults to True):
+            Whether or not to use a Fast tokenizer if possible.
+
+        token (Union[str, bool], *optional*):
+            The token to use as HTTP bearer authorization for remote files. If True, will use the token generated when
+            running `huggingface-cli login` (stored in `~/.huggingface`).
+
+        revision (str, *optional*):
+            The specific model version to use, specified as a branch name, tag name, or commit id.
```
you can remove spaces between each args

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+    if model.config is None:
+        if model_type is None:
+            raise ValueError(
+                "Could not automatically find model_type for the model, you must pass a " "model_type explictly"
```
```suggestion
                'Could not automatically find a `model_type` for the model, you must specify the `model_type` argument explicitly.'
```

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+
+        if model_type not in pipeline_map:
+            raise ValueError(
+                f"Model type: {model_type} is not supported by Ryzen pipelines. Please open a issue / PR to support the same!"
```
> to support the same

fxmarty commented on optimum/amd/ryzenai/pipelines/__init__.py:
```diff
@@ -0,0 +1,196 @@
+                        break
+                if image_processor is None:
+                    raise ValueError(
+                        "Could not automatically find image processor for the model, you must pass a "
+                        "image_processor explictly"
```
```suggestion
                        "Could not automatically find an image processor for the model, you must specifiy the argument `image_processor` explicitly."
```

fxmarty commented:
I added a few comments but it looks good.

@mht-sharma only a personal opinion, but force push & rebasing makes it harder for reviewers to understand the code changes since the last review.

mht-sharma commented:
> I added a few comments but it looks good.
> 
> @mht-sharma only a personal opinion, but force push & rebasing makes it harder for reviewers to understand the code changes since the last review (I can't know what was committed before and after the review)
> 
> https://wincent.com/wiki/git_rebase%3A_you%27re_doing_it_wrong

Sure, make keep this in mind. Thanks!
